### ISSUE TITLE
LibGfx::GIFLoader: OOB write/buffer overflow in decode_frame

### ISSUE THREAD

**Issue by Lubrsi** [SerenityOS/serenity#3484]
Found while fuzzing.

You can achieve an OOB write/buffer overflow on this line:
https://github.com/SerenityOS/serenity/blob/ad7030ce4ab53660cb12671be40f3982ea232261/Libraries/LibGfx/GIFLoader.cpp#L359

GIF:
![crash-7a7e63267ee32744ff39e0fb13953b5f87a3a151](https://user-images.githubusercontent.com/25595356/93031198-dd469c00-f620-11ea-8d1d-02aab654fcc8.gif)
(note that different GIF implementations don't completely agree on how to interpret this, e.g. on my machine it shows the green square and light green circle to the right, but on iOS, it's a black background but the coloured squares still work just fine)

Trace with process regions:
```
[FileManager(24:25)]: CPU[0] NP(error) fault at invalid address V0x013cc000
[FileManager(24:25)]: Unrecoverable page fault, write to address V0x013cc000
[FileManager(24:25)]: CRASH: CPU #0 Page Fault. Ring 3.
[FileManager(24:25)]: exception code: 0006 (isr: 0000
[FileManager(24:25)]:   pc=001b:080aea9d flags=0202
[FileManager(24:25)]:  stk=0023:013b4b90
[FileManager(24:25)]:   ds=0023 es=0023 fs=0023 gs=002b
[FileManager(24:25)]: eax=013c2000 ebx=ff8fbc8f ecx=00d72018 edx=013c2550
[FileManager(24:25)]: ebp=013b4cd8 esp=c462afe8 esi=00dd080c edi=00009ab0
[FileManager(24:25)]: cr0=80010013 cr2=013cc000 cr3=02d6f000 cr4=003006e4
[FileManager(24:25)]: PV(error) fault in Region{0xc02bf538}[102] at V0x080aea9d
WindowServer(13): Decoding a ShareableBitmap with shbuf_id=86, size=[16x16]
[FileManager(24:25)]: 0x080aea9d  Gfx::decode_frame(Gfx::GIFLoadingContext&, unsigned long) +1815

[FileManager(24:25)]: 0xc0164ac3  Kernel::Process::crash(int, unsigned int, bool) +1041
[FileManager(24:25)]: 0xc0120c30  illegal_instruction_handler +0
[FileManager(24:25)]: 0xc01222f1  page_fault_handler +1107
[FileManager(24:25)]: 0xc011e44f  page_fault_asm_entry +38
[FileManager(24:25)]: 0x080aebf1  Gfx::GIFImageDecoderPlugin::frame(unsigned long) +75
[FileManager(24:25)]: 0x080aecb2  Gfx::GIFImageDecoderPlugin::bitmap() +44
[FileManager(24:25)]: 0x080aed45  Gfx::load_gif(AK::StringView const&) +73
[FileManager(24:25)]: 0x080a8a1e  Gfx::Bitmap::load_from_file(AK::StringView const&) +272
[FileManager(24:25)]: 0x08071518  AK::Function<AK::RefPtr<Gfx::Bitmap> ()>::CallableWrapper<GUI::FileSystemModel::fetch_thumbnail_for(GUI::FileSystemModel::Node const&)::{lambda()#1}>::call() const +46
[FileManager(24:25)]: 0x080735ef  LibThread::BackgroundAction<AK::RefPtr<Gfx::Bitmap> >::BackgroundAction(AK::Function<AK::RefPtr<Gfx::Bitmap> ()>, AK::Function<void (AK::RefPtr<Gfx::Bitmap>)>)::{lambda()#1}::operator()() const +49
[FileManager(24:25)]: 0x080c9536  background_thread_func() +378
[FileManager(24:25)]: 0x080c976b  LibThread::Thread::start()::{lambda(void*)#1}::_FUN(void*) +37
[FileManager(24:25)]: 0x080c99a5  pthread_create_helper +12
[FileManager(24:25)]: Process regions:
[FileManager(24:25)]: BEGIN       END         SIZE        ACCESS  NAME
[FileManager(24:25)]: 00901000 -- 00901fff    00001000    RW        (null)
[FileManager(24:25)]: 00903000 -- 00d02fff    00400000    RW  T     Stack (Main thread)
[FileManager(24:25)]: 00d04000 -- 00d04fff    00001000    RW        Thread-specific
[FileManager(24:25)]: 00d06000 -- 00d06fff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00d10000 -- 00d1ffff    00010000    RW   P    malloc: ChunkedBlock(32)
[FileManager(24:25)]: 00d21000 -- 00d24fff    00004000    R  S      /res/fonts/Katica10.font
[FileManager(24:25)]: 00d30000 -- 00d3ffff    00010000    RW   P    malloc: ChunkedBlock(8)
[FileManager(24:25)]: 00d41000 -- 00d44fff    00004000    R  S      /res/fonts/KaticaBold10.font
[FileManager(24:25)]: 00d46000 -- 00d46fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/undo.png
[FileManager(24:25)]: 00d4a000 -- 00d4afff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/redo.png
[FileManager(24:25)]: 00d50000 -- 00d5ffff    00010000    RW   P    malloc: ChunkedBlock(16)
[FileManager(24:25)]: 00d63000 -- 00d63fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/edit-cut.png
[FileManager(24:25)]: 00d69000 -- 00d69fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/edit-copy.png
[FileManager(24:25)]: 00d70000 -- 00d7ffff    00010000    RW   P    malloc: ChunkedBlock(64)
[FileManager(24:25)]: 00d81000 -- 00d81fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/paste.png
[FileManager(24:25)]: 00d85000 -- 00d85fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/delete.png
[FileManager(24:25)]: 00d8b000 -- 00d8bfff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/select-all.png
[FileManager(24:25)]: 00d90000 -- 00d9ffff    00010000    RW   P    malloc: ChunkedBlock(128)
[FileManager(24:25)]: 00da6000 -- 00da9fff    00004000    R  S      /res/fonts/KaticaBold10.font
[FileManager(24:25)]: 00dad000 -- 00dadfff    00001000    RW   P    Gfx::Bitmap [9x9] - Decoded PNG: /res/icons/serenity/treeview-expand.png
[FileManager(24:25)]: 00db0000 -- 00dbffff    00010000    RW   P    malloc: ChunkedBlock(4090)
[FileManager(24:25)]: 00dc5000 -- 00dc5fff    00001000    RW   P    Gfx::Bitmap [9x9] - Decoded PNG: /res/icons/serenity/treeview-collapse.png
[FileManager(24:25)]: 00dcb000 -- 00dcbfff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-folder.png
[FileManager(24:25)]: 00dd0000 -- 00ddffff    00010000    RW   P    malloc: ChunkedBlock(2036)
[FileManager(24:25)]: 00de4000 -- 00de4fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-folder.png
[FileManager(24:25)]: 00de8000 -- 00de8fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-folder-open.png
[FileManager(24:25)]: 00df0000 -- 00dfffff    00010000    RW   P    malloc: ChunkedBlock(252)
[FileManager(24:25)]: 00e01000 -- 00e01fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/home-directory.png
[FileManager(24:25)]: 00e06000 -- 00e06fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/home-directory.png
[FileManager(24:25)]: 00e0a000 -- 00e0afff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/home-directory-open.png
[FileManager(24:25)]: 00e10000 -- 00e1ffff    00010000    RW   P    malloc: ChunkedBlock(1016)
[FileManager(24:25)]: 00e23000 -- 00e23fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-unknown.png
[FileManager(24:25)]: 00e2a000 -- 00e2afff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-unknown.png
[FileManager(24:25)]: 00e30000 -- 00e3ffff    00010000    RW   P    malloc: ChunkedBlock(508)
[FileManager(24:25)]: 00e43000 -- 00e43fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-symlink.png
[FileManager(24:25)]: 00e4a000 -- 00e4afff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-symlink.png
[FileManager(24:25)]: 00e50000 -- 00e5ffff    00010000    RW   P    malloc: ChunkedBlock(8188)
[FileManager(24:25)]: 00e63000 -- 00e63fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-socket.png
[FileManager(24:25)]: 00e6a000 -- 00e6afff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-socket.png
[FileManager(24:25)]: 00e70000 -- 00e7ffff    00010000         P    malloc: ChunkedBlock(32756)
[FileManager(24:25)]: 00e83000 -- 00e83fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-executable.png
[FileManager(24:25)]: 00e8b000 -- 00e8bfff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-executable.png
[FileManager(24:25)]: 00e8f000 -- 00e8ffff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-cplusplus.png
[FileManager(24:25)]: 00e96000 -- 00e96fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-cplusplus.png
[FileManager(24:25)]: 00e9a000 -- 00e9afff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-form.png
[FileManager(24:25)]: 00ea1000 -- 00ea1fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-form.png
[FileManager(24:25)]: 00ea5000 -- 00ea5fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-hackstudio.png
[FileManager(24:25)]: 00eac000 -- 00eacfff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-hackstudio.png
[FileManager(24:25)]: 00eb0000 -- 00eb0fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-header.png
[FileManager(24:25)]: 00eb7000 -- 00eb7fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-header.png
[FileManager(24:25)]: 00ebb000 -- 00ebbfff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-html.png
[FileManager(24:25)]: 00ec2000 -- 00ec2fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-html.png
[FileManager(24:25)]: 00ec9000 -- 00ec9fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-image.png
[FileManager(24:25)]: 00ed1000 -- 00ed1fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-image.png
[FileManager(24:25)]: 00ed5000 -- 00ed5fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-java.png
[FileManager(24:25)]: 00edb000 -- 00edbfff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-javascript.png
[FileManager(24:25)]: 00ee2000 -- 00ee2fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-javascript.png
[FileManager(24:25)]: 00ee6000 -- 00ee6fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-library.png
[FileManager(24:25)]: 00eed000 -- 00eedfff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-library.png
[FileManager(24:25)]: 00ef1000 -- 00ef1fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-markdown.png
[FileManager(24:25)]: 00ef8000 -- 00ef8fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-markdown.png
[FileManager(24:25)]: 00efc000 -- 00efcfff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-music.png
[FileManager(24:25)]: 00f03000 -- 00f03fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-music.png
[FileManager(24:25)]: 00f07000 -- 00f07fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-object.png
[FileManager(24:25)]: 00f0e000 -- 00f0efff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-object.png
[FileManager(24:25)]: 00f12000 -- 00f12fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-pdf.png
[FileManager(24:25)]: 00f18000 -- 00f18fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-python.png
[FileManager(24:25)]: 00f1f000 -- 00f1ffff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-python.png
[FileManager(24:25)]: 00f23000 -- 00f23fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-sound.png
[FileManager(24:25)]: 00f2a000 -- 00f2afff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-sound.png
[FileManager(24:25)]: 00f2e000 -- 00f2efff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-ini.png
[FileManager(24:25)]: 00f35000 -- 00f35fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-ini.png
[FileManager(24:25)]: 00f39000 -- 00f39fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-text.png
[FileManager(24:25)]: 00f40000 -- 00f40fff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-text.png
[FileManager(24:25)]: 00f44000 -- 00f44fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/filetype-spreadsheet.png
[FileManager(24:25)]: 00f4b000 -- 00f4bfff    00001000    RW   P    Gfx::Bitmap [32x32] - Decoded PNG: /res/icons/32x32/filetype-spreadsheet.png
[FileManager(24:25)]: 00f4d000 -- 00f4dfff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00f53000 -- 00f53fff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00f59000 -- 00f59fff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00f61000 -- 00f61fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/delete.png
[FileManager(24:25)]: 00f67000 -- 00f67fff    00001000    RW   P    Gfx::Bitmap [7x10] - Decoded PNG: /res/emoji/U+2B06.png
[FileManager(24:25)]: 00f6b000 -- 00f6bfff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00f71000 -- 00f71fff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00f77000 -- 00f77fff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00f7d000 -- 00f7dfff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00f83000 -- 00f83fff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00f89000 -- 00f89fff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00f8f000 -- 00f8ffff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00f97000 -- 00f97fff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/paste.png
[FileManager(24:25)]: 00f9b000 -- 00f9bfff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00fa1000 -- 00fa1fff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00fa7000 -- 00fa7fff    00001000    RW        Thread-specific
[FileManager(24:25)]: 00faf000 -- 00faffff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/delete.png
[FileManager(24:25)]: 00fb1000 -- 00fb1fff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00fb3000 -- 00fb3fff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 00fb5000 -- 013b4fff    00400000    RW  T     Thread stack
[FileManager(24:25)]: 013b8000 -- 013b8fff    00001000    R  S P    SharedBuffer
[FileManager(24:25)]: 013bc000 -- 013bcfff    00001000    RW   P    Gfx::Bitmap [16x16] - Decoded PNG: /res/icons/16x16/home-directory.png
[FileManager(24:25)]: 013c0000 -- 013c0fff    00001000    R  S      /home/anon/crash-7a7e63267ee32744ff39e0fb13953b5f87a3a151.gif
[FileManager(24:25)]: 013c2000 -- 013cbfff    0000a000    RW   P    GraphicsBitmap [100x100]
[FileManager(24:25)]: 013cd000 -- 013d6fff    0000a000    RW   P    GraphicsBitmap [100x100]
[FileManager(24:25)]: 013d8000 -- 01503fff    0012c000    RW S P    SharedBuffer
[FileManager(24:25)]: 08048000 -- 08121fff    000da000    R XS      elf-map-rx
[FileManager(24:25)]: 08122000 -- 08127fff    00006000    RW        elf-alloc-rw
```

**Comment by awesomekling** [comment]
cc @peterdn

**Comment by peterdn** [comment]
Nice find @Lubrsi, proposed fix in #3899.

The difference in rendering between different browsers is interesting alright! It's down to the spec not being totally clear on what to do when a frame extends beyond the limits of the global logical screen. The various behaviour appears to be:

Firefox (and now SerenityOS): don't expand logical screen, render unwritten pixels as transparent
Chrome, Edge, IE11: expand logical screen, render unwritten pixels as transparent
Safari: don't expand logical screen, render unwritten pixels as GIF background colour [this is why it's black]
